File: 3-Longest_Substring_Without_Repeating_Characters/main.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        """
        sliding window
        60ms
        """
        index = {}
        maxLength = left = right = 0
        for i, c in enumerate(s):
            right = i
            if None != index.get(c):
                if left <= index.get(c):
                    left = index.get(c)+1
            index[c] = i
            curr = right-left+1
            maxLength = curr if curr>maxLength else maxLength
        return maxLength



import unittest
class TestStringMethods(unittest.TestCase):
    obj = Solution()
    def test_1(self):
        ret = self.obj.lengthOfLongestSubstring('abcabcbb')
        self.assertEqual(ret, 3)
    def test_2(self):
        ret = self.obj.lengthOfLongestSubstring('asdfa')
        self.assertEqual(ret, 4)
    def test_3(self):
        ret = self.obj.lengthOfLongestSubstring('dvdf')
        self.assertEqual(ret, 3)
    def test_4(self):
        ret = self.obj.lengthOfLongestSubstring('aab')
        self.assertEqual(ret, 2)
    def test_5(self):
        ret = self.obj.lengthOfLongestSubstring('abba')
        self.assertEqual(ret, 2)
    def test_6(self):
        ret = self.obj.lengthOfLongestSubstring('tmmzuxt')
        self.assertEqual(ret, 5)

if __name__ == '__main__':
    unittest.main()

File: 3-Longest_Substring_Without_Repeating_Characters/main.php
<?php
class Solution {
    /**
     * sliding wiindow
     * 24ms
     * @param String $s
     * @return Integer
     */
    function lengthOfLongestSubstring($s) {
        if (empty($s)) {
            return 0;
        }
        $index = [];
        $max_length = $left = $right = 0;
        foreach (str_split($s) as $k => $c) {
            $right = $k;
            if (isset($index[$c]) && $left <= $index[$c]) {
                $left = $index[$c] + 1;
            }
            $index[$c] = $k;
            $curr = $right - $left + 1;
            $max_length = ($curr > $max_length) ? $curr : $max_length;
        }
        return $max_length;
    }
}
$obj = new Solution();
echo $obj->lengthOfLongestSubstring('abb');